spongos/src/error.rs

```diff
@@ -1,39 +1,65 @@
-use core::fmt::{Debug, Display};
+use core::{array::TryFromSliceError, fmt::Debug};
 
-use displaydoc::Display;
+use alloc::string::String;
+use thiserror_no_std::Error;
 
-#[derive(Display, Debug)]
+pub type Result<T> = core::result::Result<T, Error>;
+
+#[derive(Debug, Error)]
 pub enum Error {
     //////////
     // Generic
     //////////
-    /// Size of vec/array does not match (expected: {0}, found: {1})
+    #[error("Size of vec/array does not match (expected: {0}, found: {1})")]
     LengthMismatch(usize, usize),
 
     //////////
     // DDML Wrap/Unwrap
     //////////
-    /// There was an issue with the calculated signature, cannot unwrap message
+    #[error("here was an issue with the calculated signature, cannot unwrap message")]
     SignatureMismatch,
-    /// Failure to generate ed25519 public key
-    PublicKeyGenerationFailure,
-    /// Integrity violation. Bad MAC
+    #[error("Failure to generate ed25519 public key: {0:?}")]
+    PublicKeyGenerationFailure(crypto::Error),
+    #[error("Failed to generate slice from reference: {0:?}")]
+    SliceMismatch(TryFromSliceError),
+    #[error("Integrity violation. Bad MAC")]
     BadMac,
 
+    #[error("Failed to \"{0}\" for {1}. Reason: {2}")]
+    InvalidAction(&'static str, String, String),
+
+    #[error("{1} is not a valid {0} option")]
+    InvalidOption(&'static str, u8),
+
+    #[error("{0} version '{1}' not supported")]
+    Version(&'static str, u8),
+
+    #[error("Reserved area was not empty: {0}")]
+    Reserved(&'static str),
+
     //////////
     // DDML IO
     //////////
-    /// Not enough space allocated for output stream (expected: {0}, found: {1})
+    #[error("Not enough space allocated for output stream (expected: {0}, found: {1})")]
     StreamAllocationExceededOut(usize, usize),
-    /// Not enough space allocated for input stream (expected: {0}, found: {1})
+    #[error("Not enough space allocated for input stream (expected: {0}, found: {1})")]
     StreamAllocationExceededIn(usize, usize),
+
+    #[error("Generating context {0} failed due to {1}")]
```

Comment:
I think this is the most confusing error to understand as is. In most implementations it reads like this: 
`Generating context ContentVerify failed due to [error]` which doesn't really make any sense. We should probably rephrase it to something along the lines of 

 ```suggestion
    #[error("Context failed to perform the message command \"{0}\"; Error: {1}")]
```

This way it would read `Context failed to perform the message command "ContentVerify"; Error: [error]`

Reply:
Yea this one was a hard one, as it feels unneeded but needs to be thrown from the spongos error throughout streams/lets